raft-tech/TANF-app
# Issue 1013: Hotfix part 2

### Summary of Changes
Closes #1013 

Resolves functional issues with the previous implementation of 1013 by updating the GitHub Action to only trigger off the `QASP Review` label and instead trigger a separate workflow for `owasp-scan` in Circle CI. This PR also removes the OWASP scan steps from the pre-existing `test-backend` and `test-frontend` jobs. Now, those steps will only run for PRs in QASP Review.

Additional changes:
* Added reusable commands to Circle CI config - `docker-compose-up-backend`, `docker-compose-up-frontend` and `run-owasp-scan`.
* Removed parameters from `test-backend` and utilize new docker-compose commands
* Removed parameters from `test-frontend`
* Added jobs for `backend-owasp-scan` and `frontend-owasp-scan`
* Update job `nightly-owasp-scan` to use the new `run-owasp-scan` command
* Renamed GH Action workflow `check-on-label.yml` to `qasp-owasp-scan.yml` and added concurrency controls.
* Updated documentation
* Miscellaneous cleanup and pipeline parameter naming consistency improvements

### How to Test
This is a DevOps only change. Testing will be performed by assigning labels to this PR and making test commits.

Resources needed:
- [Github Actions workflow results](https://github.com/raft-tech/TANF-app/actions/workflows/qasp-owasp-scan.yml)
- [This branch in Circle CI](https://app.circleci.com/pipelines/github/raft-tech/TANF-app?branch=hotfix%2F1013-part-2)

1. Assign a random label to this PR (like backend, frontend, etc) and ensure that the GitHub Action is skipped.
2. Assign the `QASP Review` label to this PR and ensure the GitHub Action is triggered and the `owasp-scan` workflow is initiated in Circle CI. Also ensure that `build-and-test` is not re-run when applying the label.
3. Assign a different label to this PR and ensure that the `owasp-scan` workflow is not triggered a second time.
4. Make a commit to this branch (add something random to docs, etc) and ensure that both the `build-and-test` and `owasp-scan` workflows are triggered independently in Circle CI.
5. Remove the `QASP Review` label and ensure that no Circle CI pipelines were triggered.
6. Make a new commit undoing the change from step 4 and ensure that only `build-and-test` is run.

> Expected workflow output for PRs with `QASP Review` label:
![Screenshot from 2021-11-02 13-31-53](https://user-images.githubusercontent.com/22626085/139916229-cf8d6639-f2e0-4a93-810a-d05ed7454bb1.png)

### Deliverable 1: Accepted Features

> *Performance Standard(s): At the beginning of each sprint, the Product Owner and development team will collaborate to define a set of user stories to be completed during the sprint. Acceptance criteria for each story will also be defined. The development team will deliver code and functionality to satisfy these user stories.*

> *Acceptable Quality Level: Delivered code meets the acceptance criteria for each user story. Incomplete stories will be assessed and considered for inclusion in the next sprint.*

  - [x] OWASP scan does not run with build-and-test workflow by default
  - [x] QASP Review label causes CircleCI to *not* skip jobs and run every job like we do now.
  - [x] Ensure [documentation](https://github.com/raft-tech/TANF-app/blob/raft-tdp-main/docs/Technical-Documentation/circle-ci.md) is up to date with latest code.
  - [x] reviewed by @ADPennington 

As Product Owner, @lfrohlich will decide if ACs are met.

### Deliverable 2: Tested Code

> *Performance Standard(s): Code delivered under the order must have substantial test code coverage. Version-controlled HHS GitHub repository of code that comprises products that will remain in the government domain.*

> *Acceptable Quality Level: Minimum of 90% test coverage of all code. All areas of code are meaningfully tested.*

+ Are all areas of code introduced in this PR meaningfully tested?
+ Are code coverage minimums met?

### Deliverable 3: Properly Styled Code

> *Performance Standard(s): GSA 18F Front- End Guide*

> *Acceptable Quality Level: 0 linting errors and 0 warnings*


### Deliverable 4: Accessible

> *Performance Standard(s): Web Content Accessibility Guidelines 2.1 AA standards*

> *Acceptable Quality Level: 0 errors reported using an automated scanner and 0 errors reported in manual testing*
    + [See the full Accessibility Assessment plan here.](https://github.com/HHS/TANF-app/blob/main/docs/Technical-Documentation/how-government-will-test-a11y.md)

_ci changes only_

### Deliverable 5: Deployed

> *Performance Standard(s): Code must successfully build and deploy into the staging environment.*

> *Acceptable Quality Level: Successful deployment by assigning a 'Deploy with CircleCI - {env}' label*


### Deliverable 6: Documented

> *Performance Standard(s): Summary of user stories completed every two weeks. All dependencies are listed and the licenses are documented. Major functionality in the software/source code is documented, including system diagram. Individual methods are documented inline in a format that permits the use of tools such as JSDoc. All non-inherited 800-53 system security controls are documented in the Open Control or OSCAL format and HHS Section 508 Product Assessment Template (PAT) are updated as appropriate.*

> *Acceptable Quality Level: Combination of manual review and automated testing, if available*


_ci changes only_

### Deliverable 7: Secure

> *Performance Standard(s): Open Web Application Security Project (OWASP) Application Security Verification Standard 3.0*

> *Acceptable Quality Level: Code submitted must be free of medium- and high-level static and dynamic security vulnerabilities*


_no new issues detected_

### Deliverable 8: Context

> * Performance Standard(s): Code must be understandable and contextualized for the reviewers possess the knowledge and background necessary for analysis and constructive criticism to take place.

> * Acceptable Quality Level: Code submitted in the pull request has context.*
